### PR TITLE
add support for disabling data table selection (checkbox/radio)

### PR DESCRIPTION
**Pull request by meizibupt**
Fixes #[3125](https://github.com/salesforce/design-system-react/issues/3125)

### Additional description
add option to disable radio/checkbox for datatable
image
image

---

### CONTRIBUTOR checklist (do not remove)
Please complete for every pull request

* [ ] First-time contributors should sign the Contributor License Agreement. It's a fancy way of saying that you are giving away your contribution to this project. If you haven't before, wait a few minutes and a bot will comment on this pull request with instructions.
* [x] `npm run lint:fix` has been run and linting passes.
* [x] Mocha, Jest (Storyshots), and `components/component-docs.json` CI tests pass (`npm test`).
* [x] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [x] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [x] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).

### REVIEWER checklist (do not remove)

* [ ] CircleCI tests pass. This includes linting, Mocha, Jest, Storyshots, and `components/component-docs.json` tests.
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).
###### Required only if there are markup / UX changes
* [ ] Add year-first date and commit SHA to `last-slds-markup-review` in `package.json` and push.
* [ ] Request a review of the deployed Heroku app by the Salesforce UX Accessibility Team.
* [ ] Add year-first review date, and commit SHA, `last-accessibility-review`, to `package.json` and push.
* [ ] While the contributor's branch is checked out, run `npm run local-update` within locally cloned [site repo](https://github.com/salesforce-ux/design-system-react-site) to confirm the site will function correctly at the next release.

### CODE REVIEW

**Comment by interactivellama** [comment]
@meizibupt I'm going to look at this a little more later tonight, but when I pulled it down and used it. The main issue I'm seeing is that Nav to Edit mode and try to set focus on the disabled item. To reproduce, navigate with the arrow keys and then press enter. Focus visually disappears but it looks like the state is still there. 

I would recommend disabling switching to edit mode, essentially ignoring the switch to edit mode

**Comment by interactivellama** [comment]
Demo:

**Comment by meizibupt** [comment]
> @meizibupt I'm going to look at this a little more later tonight, but when I pulled it down and used it. The main issue I'm seeing is that Nav to Edit mode and try to set focus on the disabled item. To reproduce, navigate with the arrow keys and then press enter. Focus visually disappears/gets lost.
> 
> I would recommend disabling switching to edit mode, essentially ignoring the switch to edit mode

what do you mean by edit mode? do you mean disallowing navigating to the disabled checkbox?

**Comment by interactivellama** [comment]
Yes, as in ignore the enter key if the child is disabled which is what is done when the content is only text in a cell.

Please check me on this, but I believe returning a normal disabled `<Checkbox>` instead of `<InteractiveCheckbox>` here https://github.com/salesforce/design-system-react/blob/master/components/data-table/private/row.jsx#L145
would disable the focus and treat it as non-actionable / non-editable.

Same with `Radio` here https://github.com/salesforce/design-system-react/blob/master/components/data-table/private/row.jsx#L145

File: components/data-table/__examples__/advanced-single-select-fixed-header.jsx [hunk 1]
```diff
@@ -124,7 +130,8 @@ class Example extends React.Component {
 						onRowChange={this.handleChanged}
 						onSort={this.handleSort}
 						selection={this.state.selection}
-						selectRows="radio"
+						disabledSelection={this.state.disabledSelection}
+						selectRows="checkbox"
```
Comment (interactivellama) [inline components/data-table/__examples__/advanced-single-select-fixed-header.jsx:134]: This is the single select example for consumers. Please keep `radio` or choose another example.

File: components/data-table/__examples__/advanced-single-select-fixed-header.jsx [hunk 2]
```diff
@@ -23,43 +56,16 @@ class Example extends React.Component {
...
-				amount: '$25,000',
-				contact: 'nathan@salesforce.com',
-			},
-		],
+		items,
 		selection: [],
+		disabledSelection: [items[0]],
 	};
 
 	handleChanged = (event, data) => {
```
Comment (interactivellama) [inline components/data-table/__examples__/advanced-single-select-fixed-header.jsx:64]: Does the event fire on a disabled checkbox?

File: components/data-table/__tests__/data-table.browser-test.jsx [hunk 3]
```diff
@@ -248,6 +248,26 @@ describe('DataTable: ', function describeFunction() {
...
+						<DataTableColumn {...columnProps} key={columnProps.property} />
+					))}
+				</DataTable>
+			).call(this);
+
+			const tbody = getTable(this.dom).querySelectorAll('tbody')[0];
+			const disabledRows = tbody.querySelectorAll(
+				'.slds-checkbox input:disabled'
+			);
+			disabledRows.should.have.length(1);
```
Comment (interactivellama) [inline components/data-table/__tests__/data-table.browser-test.jsx:268]: We should spy the `onRowChange` and make sure it doesn't fire OR make sure the data.selection parameter is still `0` in length. See `can select a row` below and negate it.

File: components/data-table/__examples__/advanced-single-select-fixed-header.jsx (hunk 2 above)
Comment (interactivellama) [reply components/data-table/__examples__/advanced-single-select-fixed-header.jsx:64]: I'm asking about this line. 

I'm thinking that disabled doesn't mean a row can't be selected. You could force select a row (that is make it selected) and then disable it. 

```
const selections = data.selection.filter(
			(s) => !this.state.disabledSelection.includes(s)
		);
```

File: components/data-table/__tests__/data-table.browser-test.jsx (hunk 3 above)
Comment (meizibupt) [reply components/data-table/__tests__/data-table.browser-test.jsx:271]: updated test

File: components/data-table/__examples__/advanced-single-select-fixed-header.jsx (hunk 1 above)
Comment (meizibupt) [reply components/data-table/__examples__/advanced-single-select-fixed-header.jsx]: reverted the changes.  I added a disabled checkbox in fixed-header example, which is showing checkbox currently

File: components/data-table/__examples__/advanced-single-select-fixed-header.jsx (hunk 2 above)
Comment (meizibupt) [reply components/data-table/__examples__/advanced-single-select-fixed-header.jsx]: yes. Makes sense.  A row can be preselected but disabled.

**Review by interactivellama — APPROVED** [review]

**Comment by interactivellama** [comment]
Great addition! @meizibupt

**Pull request merged**